src/platform/backends/hyperv_api/hcs_virtual_machine.cpp

```diff
@@ -0,0 +1,748 @@
+
+    switch (type)
+    {
+    case hcs::HcsEventType::SystemExited:
+    {
+        mpl::info(log_category,
+                  "compute_system_event_callback() > {}:  SystemExited event received",
+                  vm->get_name());
+        vm->state = State::off;
+        vm->handle_state_update();
+    }
+    break;
+    case hcs::HcsEventType::Unknown:
+        mpl::info(log_category,
+                  "compute_system_event_callback() > {}:  Unidentified event received",
+                  vm->get_name());
+        break;
+    }
+}
+
+std::filesystem::path HCSVirtualMachine::get_primary_disk_path() const
+{
+    const std::filesystem::path base_vhdx = description.image.image_path.toStdString();
+    const std::filesystem::path head_avhdx =
+        base_vhdx.parent_path() / virtdisk::VirtDiskSnapshot::head_disk_name();
+    return std::filesystem::exists(head_avhdx) ? head_avhdx : base_vhdx;
+}
+
+void HCSVirtualMachine::grant_access_to_paths(std::list<std::filesystem::path> paths) const
+{
+    // std::list, because we need iterator and pointer stability while inserting.
+    // Normal for loop here because we want .end() to be evaluated in every
+    // iteration since we might also insert new elements to the list.
+    for (auto itr = paths.begin(); itr != paths.end(); ++itr)
+    {
+        const auto& path = *itr;
+        mpl::debug(log_category,
+                   "Granting access to path `{}`, exists? {}",
+                   path,
+                   std::filesystem::exists(path));
+        if (std::filesystem::is_symlink(path))
+        {
+            paths.push_back(std::filesystem::canonical(path));
+        }
+
+        if (const auto r = HCS().grant_vm_access(get_name(), path); !r)
+        {
+            mpl::error(log_category,
+                       "Could not grant access to VM `{}` for the path `{}`, error code: {}",
+                       get_name(),
+                       path,
+                       r);
+        }
+    }
+}
+
+bool HCSVirtualMachine::maybe_create_compute_system()
+{
+    // Always reset the handle and create a new one.
+    hcs_system.reset();
+    auto attach_callback_handler = sg::make_scope_guard([this]() noexcept {
+        if (hcs_system)
+        {
+            top_catch_all(log_category, [this] {
+                if (!HCS().set_compute_system_callback(
+                        hcs_system,
+                        this,
+                        HCSVirtualMachine::compute_system_event_callback))
+                {
+                    mpl::warn(log_category,
+                              "Could not set compute system callback for VM: `{}`!",
+                              get_name());
+                }
+            });
+        }
+    });
+
+    const auto result = HCS().open_compute_system(get_name(), hcs_system);
+    if (result)
+    {
+        // Opened existing VM
+        return false;
+    }
+
+    // FIXME: Handle suspend state?
```

Comment:
What's missing for this?

Reply:
It should be trivial to implement. I'll implement it after the christmas break.